### ISSUE TITLE
toString function is executing function expression assign as a value to an "equals" key of given object.

### ISSUE THREAD

**Issue by ValkaHonda** [ramda/ramda#3150]
The **Bug** is reproduced [here](https://ramdajs.com/repl/?v=0.27.0#?const%20complexObject%20%3D%20%7B%20%0A%20%20foo%3A%20%7B%20%0A%20%20%20%20foo%3A%20%7B%0A%20%20%20%20%20%20equals%3A%20%28%29%20%3D%3E%20console.log%28%27This%20should%20NOT%20be%20called%2C%20but%20it%20is%21%20D%3A%27%29%20%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%0Aconst%20simplerObject%20%3D%20%7B%20%0A%20%20foo%3A%20%7B%0A%20%20%20%20equals%3A%20%28%29%20%3D%3E%20console.log%28%27This%20will%20not%20be%20called%27%29%20%0A%20%20%7D%0A%7D%0A%0AR.toString%28complexObject%29%0A)!

![image](https://user-images.githubusercontent.com/26555796/119011382-b0ac0680-b99d-11eb-9f5e-50e0bfee298e.png)

You can see that the **toString** function will execute the function, which is provided on the **equals** property(**please** see *complexObject*).
Also **Note** how I have two nested identical keys(in this case -> **foo**).

If I change the **equals** property to something different then there is **no problem**. If I provide **simplerObject** to the **toString** function, then equals is **NOT** being executed as well.

I would expect the **toString** function to **never** execute functions the way it is shown on the above example, which is showing the **Bug**. 

What is the source of this **Bug and can it be fixed?

**Comment by CrossEye** [comment]
I can tell you briefly that this certainly has to do with cycle-detection.  It's pretty hard to pretty-print something like this:

```js
const x = {}
x.x = x

toString(x)
```

That returns `{"x": <Circular>}`, which is about the best we can do.  But in testing equality, we delegate to an object's `equals` method if it has one.

This is definitely a bug.  But I don't immediately see a good fix.  The best I can imagine is to tell the internal `equals` call not to delegate in certain circumstances.  But that feels nasty.  If you want to take a look and come up with suggestions, I'm all ears.

This is the relevant chain of calls:

  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/toString.js#L41
  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/internal/_toString.js#L12
  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/internal/_includes.js#L5
  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/internal/_indexOf.js#L51
  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/equals.js#L31 
  - https://github.com/ramda/ramda/blob/6b6a85d3fe30ac1a41ac05734be9f61bd92325e5/source/internal/_equals.js#L50-L53

That penultimate step is silly.  The internal `_indexOf` should be directly calling the internal `_equals` directly, and not going through the public `equals` call.  But that has nothing to do with your issue.

**Comment by boyskila** [comment]
For those who are using toString  with memoizeWith to generate a key, possible workaround would be just to rename the passed object `equals` . Some pseudo code:
`
const memo = R.memoizeWith(R.toString,  fn)
const renameEquals = renameProp('equals')
.............
const memoized = R.pipe(
  renameEquals('eq')
  memo,
)

const memoizedObj = memoized(obj)
..............
`

I now It's far from the greatest solution (and we will be really happy if someone can provide a better one) but it is working at least.

**Comment by CrossEye** [comment]
I'm thinking that instead of using the internal `_includes` -- which then calls `equals` -- for testing circularity, we should instead just test reference equality.  I think that makes sense, but I haven't given it very deep thought yet.

Anyone have an objection?  @ValkaHonda? @ramda/core?

**Comment by customcommander** [comment]
I guess the intention with `toString` was to offer something similar to `show` in Haskell?

I'm not quite sure I understand why we need to dispatch to an `equals` method of an object if there is one. Is this is so that we can detect circular references?

There is no way we can reproduce a 100% identical data structure in all cases so could we not simply stick with serialisable data and ignore functions, DOM stuff and symbols?

**Comment by CrossEye** [comment]
@customcommander: Yes, that was the point of `toString`.  The fact that we dispatch to `equals` is simply a side-effect of the implementation details.  We can certainly work around this.

**Comment by ValkaHonda** [comment]
> I'm thinking that instead of using the internal `_includes` -- which then calls `equals` -- for testing circularity, we should instead just test reference equality. I think that makes sense, but I haven't given it very deep thought yet.
> 
> Anyone have an objection? @ValkaHonda? @ramda/core?

This make sense. @CrossEye do you see any potential problem if **reference equality** is used to work around this bug and **not** call the *equals* method?

**Comment by CrossEye** [comment]
I can't think of any issues.  I think it won't be too difficult to change.
